File: src/tools/finance/kalshi-vol-signals.ts
type KalshiEventType = 'fomc' | 'cpi' | 'nfp' | 'other';
type FetchLike = (input: string | URL | Request, init?: RequestInit) => Promise<Response>;

export interface KalshiVolSignal {
  /** ISO timestamp of the event (UTC). */
  eventAt: string;
  /** Short label, e.g. "FOMC-2026-06" or "CPI-2026-05". */
  eventId: string;
  /** Implied probability of the selected high-impact macro outcome. */
  probability: number;
  /** Suggested volatility multiplier for the days leading into the event. */
  intensityBoost: number;
  /** Classified event bucket for downstream covariate routing. */
  eventType: KalshiEventType;
  /** Original title for diagnostics / auditability. */
  sourceTitle: string;
}

export interface KalshiFetchOptions {
  /** ISO YYYY-MM-DD inclusive lower bound for events to consider. */
  fromDate: string;
  /** ISO YYYY-MM-DD inclusive upper bound for events to consider. */
  toDate: string;
  /** Override the API base URL (for testing). Default: https://trading-api.kalshi.com/trade-api/v2 */
  baseUrl?: string;
  /** Override the API key. Falls back to process.env.KALSHI_API_KEY. */
  apiKey?: string;
  /** Injectable fetch for tests. Defaults to global fetch. */
  fetchImpl?: FetchLike;
}

export interface KalshiVolatilityCovariate {
  dates: string[];
  values: number[];
  activeSignals: number;
  peakValue: number;
}

interface KalshiMarketRecord {
  ticker?: string;
  title?: string;
  subtitle?: string;
  status?: string;
  expiration_time?: string;
  close_time?: string;
  settlement_time?: string;
  yes_ask?: number;
  last_price?: number;
}

const MS_PER_DAY = 24 * 60 * 60 * 1000;

const EVENT_PATTERNS: Array<{ type: KalshiEventType; regex: RegExp; weight: number }> = [
  { type: 'fomc', regex: /\b(fomc|fed|federal reserve|rate decision|rate hike|rate cut)\b/i, weight: 1.5 },
  { type: 'cpi', regex: /\b(cpi|inflation|consumer price index)\b/i, weight: 1.3 },
  { type: 'nfp', regex: /\b(nonfarm payroll|non-farm payroll|payrolls|jobs report|nfp)\b/i, weight: 1.2 },
];

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

function parseDate(value: string): number {
  const millis = Date.parse(value);
  if (!Number.isFinite(millis)) throw new Error(`invalid Kalshi date: ${value}`);
  return millis;
}

function normalizeProbability(market: KalshiMarketRecord): number | null {
  const raw = market.yes_ask ?? market.last_price;
  if (raw === undefined || !Number.isFinite(raw)) return null;
  const probability = raw > 1 ? raw / 100 : raw;
  return probability >= 0 && probability <= 1 ? probability : null;
}

function classifyEventType(title: string): { type: KalshiEventType; weight: number } | null {
  for (const pattern of EVENT_PATTERNS) {
    if (pattern.regex.test(title)) return { type: pattern.type, weight: pattern.weight };
  }
  return null;
}

function computeIntensityBoost(probability: number, weight: number, eventAt: string, fromDate: string): number {
  const daysUntilEvent = Math.max(0, (parseDate(eventAt) - parseDate(`${fromDate}T00:00:00Z`)) / MS_PER_DAY);
  const urgency = clamp(1 - daysUntilEvent / 14, 0.25, 1);
  return clamp(0.2 + probability * weight * urgency, 0.1, 2.5);
}

export function extractKalshiVolSignalsFromPayload(
  payload: unknown,
  opts: Pick<KalshiFetchOptions, 'fromDate' | 'toDate'>,
): KalshiVolSignal[] {
  if (!payload || typeof payload !== 'object' || !Array.isArray((payload as { markets?: unknown[] }).markets)) {
    throw new Error('Kalshi response missing markets array');
  }
  const fromMs = parseDate(`${opts.fromDate}T00:00:00Z`);
  const toMs = parseDate(`${opts.toDate}T23:59:59Z`);

  return (payload as { markets: KalshiMarketRecord[] }).markets.flatMap((market) => {
    const title = market.title?.trim();
    const eventAt = market.expiration_time ?? market.close_time ?? market.settlement_time;
    if (!title || !eventAt) return [];
    const classified = classifyEventType(title);
    if (!classified) return [];
    const probability = normalizeProbability(market);
    if (probability === null) return [];
    const eventMs = parseDate(eventAt);
    if (eventMs < fromMs || eventMs > toMs) return [];
    return [{
      eventAt,
      eventId: market.ticker ?? `${classified.type}-${eventAt.slice(0, 10)}`,
      probability,
      intensityBoost: computeIntensityBoost(probability, classified.weight, eventAt, opts.fromDate),
      eventType: classified.type,
      sourceTitle: title,
    }];
  });
}

export class KalshiUnconfiguredError extends Error {
  constructor() {
    super(
      'Kalshi macro signal loader is not configured.\n' +
      '  Set KALSHI_API_KEY in the environment or pass apiKey explicitly to\n' +
      '  fetchKalshiVolSignals().',
    );
    this.name = 'KalshiUnconfiguredError';
  }
}

export async function fetchKalshiVolSignals(opts: KalshiFetchOptions): Promise<KalshiVolSignal[]> {
  const apiKey = opts.apiKey ?? process.env.KALSHI_API_KEY;
  if (!apiKey) throw new KalshiUnconfiguredError();

  const fetchImpl = opts.fetchImpl ?? fetch;
  const baseUrl = (opts.baseUrl ?? 'https://trading-api.kalshi.com/trade-api/v2').replace(/\/$/, '');
  const url = `${baseUrl}/markets?status=open&limit=200`;
  const response = await fetchImpl(url, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
      Accept: 'application/json',
    },
  });
  if (!response.ok) {
    throw new Error(`Kalshi request failed: ${response.status} ${response.statusText}`);
  }
  const payload = await response.json();
  return extractKalshiVolSignalsFromPayload(payload, opts);
}

export function buildKalshiVolatilityCovariate(
  dates: string[],
  signals: KalshiVolSignal[],
  lookaheadDays = 5,
): KalshiVolatilityCovariate {
  const values = dates.map((date) => {
    const dateMs = parseDate(`${date}T00:00:00Z`);
    let total = 0;
    for (const signal of signals) {
      const daysAhead = (parseDate(signal.eventAt) - dateMs) / MS_PER_DAY;
      if (daysAhead < 0 || daysAhead > lookaheadDays) continue;
      const decay = 1 - daysAhead / Math.max(1, lookaheadDays);
      total += signal.intensityBoost * decay;
    }
    return Number(total.toFixed(6));
  });
  return {
    dates: [...dates],
    values,
    activeSignals: signals.length,
    peakValue: values.length > 0 ? Math.max(...values) : 0,
  };
}

File: research/utils/kalshi_vol_signals.py
"""Kalshi macro volatility signals — Python mirror of the TypeScript helper."""

from __future__ import annotations

import inspect
import os
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Literal

import requests

KalshiEventType = Literal["fomc", "cpi", "nfp", "other"]
FetchLike = Callable[[str, dict[str, str]], Any | Awaitable[Any]]
MS_PER_DAY = 24 * 60 * 60 * 1000


class KalshiUnconfiguredError(RuntimeError):
    """Raised when KALSHI_API_KEY is not set."""


@dataclass(frozen=True)
class KalshiVolSignal:
    event_at: str
    event_id: str
    probability: float
    intensity_boost: float
    event_type: KalshiEventType
    source_title: str


@dataclass(frozen=True)
class KalshiVolatilityCovariate:
    dates: list[str]
    values: list[float]
    active_signals: int
    peak_value: float


EVENT_PATTERNS: tuple[tuple[KalshiEventType, re.Pattern[str], float], ...] = (
    ("fomc", re.compile(r"\b(fomc|fed|federal reserve|rate decision|rate hike|rate cut)\b", re.I), 1.5),
    ("cpi", re.compile(r"\b(cpi|inflation|consumer price index)\b", re.I), 1.3),
    ("nfp", re.compile(r"\b(nonfarm payroll|non-farm payroll|payrolls|jobs report|nfp)\b", re.I), 1.2),
)


def _parse_date(value: str) -> int:
    iso = value.replace("Z", "+00:00")
    dt = datetime.fromisoformat(iso)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return int(dt.timestamp() * 1000)


def _clamp(value: float, lower: float, upper: float) -> float:
    return min(upper, max(lower, value))


def _normalize_probability(market: dict[str, Any]) -> float | None:
    raw = market.get("yes_ask", market.get("last_price"))
    if not isinstance(raw, (int, float)):
        return None
    probability = raw / 100 if raw > 1 else float(raw)
    if 0 <= probability <= 1:
        return probability
    return None


def _classify_event_type(title: str) -> tuple[KalshiEventType, float] | None:
    for event_type, pattern, weight in EVENT_PATTERNS:
        if pattern.search(title):
            return event_type, weight
    return None


def _compute_intensity_boost(probability: float, weight: float, event_at: str, from_date: str) -> float:
    days_until_event = max(0.0, (_parse_date(event_at) - _parse_date(f"{from_date}T00:00:00Z")) / MS_PER_DAY)
    urgency = _clamp(1 - days_until_event / 14, 0.25, 1.0)
    return _clamp(0.2 + probability * weight * urgency, 0.1, 2.5)


def extract_kalshi_vol_signals_from_payload(
    payload: Any,
    *,
    from_date: str,
    to_date: str,
) -> list[KalshiVolSignal]:
    if not isinstance(payload, dict) or not isinstance(payload.get("markets"), list):
        raise ValueError("Kalshi response missing markets array")

    from_ms = _parse_date(f"{from_date}T00:00:00Z")
    to_ms = _parse_date(f"{to_date}T23:59:59Z")
    out: list[KalshiVolSignal] = []

    for market in payload["markets"]:
        if not isinstance(market, dict):
            continue
        title = market.get("title")
        event_at = market.get("expiration_time") or market.get("close_time") or market.get("settlement_time")
        if not isinstance(title, str) or not title.strip() or not isinstance(event_at, str):
            continue
        classified = _classify_event_type(title.strip())
        if classified is None:
            continue
        probability = _normalize_probability(market)
        if probability is None:
            continue
        event_ms = _parse_date(event_at)
        if event_ms < from_ms or event_ms > to_ms:
            continue
        event_type, weight = classified
        out.append(
            KalshiVolSignal(
                event_at=event_at,
                event_id=str(market.get("ticker") or f"{event_type}-{event_at[:10]}"),
                probability=probability,
                intensity_boost=_compute_intensity_boost(probability, weight, event_at, from_date),
                event_type=event_type,
                source_title=title.strip(),
            )
        )
    return out


async def fetch_kalshi_vol_signals(
    *,
    from_date: str,
    to_date: str,
    base_url: str = "https://trading-api.kalshi.com/trade-api/v2",
    api_key: str | None = None,
    fetch_impl: FetchLike | None = None,
) -> list[KalshiVolSignal]:
    key = api_key or os.environ.get("KALSHI_API_KEY")
    if not key:
        raise KalshiUnconfiguredError(
            "fetch_kalshi_vol_signals requires a KALSHI_API_KEY environment variable "
            "or an explicit api_key argument."
        )

    url = f"{base_url.rstrip('/')}/markets?status=open&limit=200"
    headers = {"Authorization": f"Bearer {key}", "Accept": "application/json"}

    if fetch_impl is None:
        response = requests.get(url, headers=headers, timeout=30)
        response.raise_for_status()
        payload = response.json()
    else:
        payload = fetch_impl(url, headers)
        if inspect.isawaitable(payload):
            payload = await payload

    return extract_kalshi_vol_signals_from_payload(payload, from_date=from_date, to_date=to_date)


def build_kalshi_volatility_covariate(
    dates: list[str],
    signals: list[KalshiVolSignal],
    lookahead_days: int = 5,
) -> KalshiVolatilityCovariate:
    values: list[float] = []
    for date in dates:
        date_ms = _parse_date(f"{date}T00:00:00Z")
        total = 0.0
        for signal in signals:
            days_ahead = (_parse_date(signal.event_at) - date_ms) / MS_PER_DAY
            if days_ahead < 0 or days_ahead > lookahead_days:
                continue
            decay = 1 - days_ahead / max(1, lookahead_days)
            total += signal.intensity_boost * decay
        values.append(round(total, 6))

    return KalshiVolatilityCovariate(
        dates=list(dates),
        values=values,
        active_signals=len(signals),
        peak_value=max(values) if values else 0.0,
    )
